### PR TITLE
[FLOC-1252] Display marketing release version in documentation releases.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1252

I decided that we should tag documentation releases (with `+doc.X`). This is stripped in the documentation.

### CODE REVIEW

File: flocker/docs/_version.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    )
+
+
+class UnparseableVersion(Exception):
+    """
+    A version was passed that was unable to be parsed.
+    """
+
+
+def parse_version(version):
```
Comment (adamtheturtle) [inline flocker/docs/_version.py]: This seems like it would be nicer if, instead of a dictionary, there was a Version object, and characteristic did the comparison work.

File: flocker/docs/_version.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    :raises UnparseableVersion: If the version can't be parsed as a flocker
+        version.
+    """
+    match = _VERSION_RE.match(version)
+    if match is None:
+        raise UnparseableVersion(version)
+    return match.groupdict()
+
+
+def get_doc_version(version):
```
Comment (adamtheturtle) [inline flocker/docs/_version.py]: Docstring

File: flocker/docs/_version.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    parts = parse_version(version)
+    if is_release(version) and parts['doc'] is not None:
+        return parts['release']
+    else:
+        return version
+
+
+def is_release(version):
```
Comment (adamtheturtle) [inline flocker/docs/_version.py]: Docstring

File: flocker/docs/test/test_version.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    def test_doc(self):
+        """
+        When the documentation version is from a doc release, it is a release.
+        """
+        self.assertTrue(is_release('0.3.2+doc.11'))
+
+    def test_doc_dirty(self):
+        """
+        When the version is from a doc release but is dirty, it isn't a
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: :question: Change 'doc release' occurrences to 'documentation release'.

**Comment by adamtheturtle** [comment]
This code should go into master too, right?

**Comment by tomprince** [comment]
> This code should go into master too, right?
https://clusterhq.atlassian.net/browse/FLOC-1282

File: flocker/docs/_version.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    :return dict: with the following keys. If a key isn't relevant, the value
+        is ``None``.
+
+        release
+            The base release
+        development
+            For development versions, the number of commits since the last
+            release and the commit hash, in the format used by ``git describe`.
+        doc
+            For a documentation only release, the relase number.
```
Comment (adamtheturtle) [inline flocker/docs/_version.py]: typo s/relase/release

**Comment by adamtheturtle** [comment]
Sorry to miss that - and this might be another, it is quite hard to follow all of these sub-tasks. Can you please confirm that there is (planned) work to update BuildBot's regexes which check for example whether this branch is a release (and so whether to merge master in) to enable this new release branch style. That is (maybe not limited to) `_RELEASE_TAG_RE` in build.clusterhq.com/flocker_bb/steps.py.

File: docs/gettinginvolved/infrastructure/release-schedule-versioning.rst [hunk 6]
```diff
@@ -53,6 +53,15 @@ Minor marketing releases will be planned and scheduled by ClusterHQ's product te
 
 The version of a minor marketing release will have the micro version number incremented from the previous marketing release.
 
+Documentation Release
+^^^^^^^^^^^^^^^^^^^^^
+Documentation releases will be made when documentation for a major or minor marketing release is to be updated, without doing a full release.
+
+Documentation releases will be planned and scheduled by ClusterHQ's product team, in consultation with the marketing and engineering teams.
+
+The version of a documentation will have the version of the corresponding marketing release, with a ``+doc.X`` release, where ``X`` starts at ``1`` and is incremented for each pre-release.
```
Comment (adamtheturtle) [inline docs/gettinginvolved/infrastructure/release-schedule-versioning.rst]: "incremented for each pre-release" > "incremented for each documentation release".

File: docs/gettinginvolved/infrastructure/release-schedule-versioning.rst [hunk 7]
```diff
@@ -53,6 +53,15 @@ Minor marketing releases will be planned and scheduled by ClusterHQ's product te
 
 The version of a minor marketing release will have the micro version number incremented from the previous marketing release.
 
+Documentation Release
+^^^^^^^^^^^^^^^^^^^^^
```
Comment (adamtheturtle) [inline docs/gettinginvolved/infrastructure/release-schedule-versioning.rst]: Can there be a prerelease of a documentation release?

Comment (tomprince) [inline docs/gettinginvolved/infrastructure/release-schedule-versioning.rst]: I'm going to say no.

File: flocker/docs/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+from ._version import get_doc_version, is_release
+
+__all__ = ['get_doc_version', is_release]
```
Comment (adamtheturtle) [inline flocker/docs/__init__.py]: This could be my lack of understanding but why is one a string and one not?

Comment (tomprince) [inline flocker/docs/__init__.py]: They should both be strings.

File: docs/gettinginvolved/infrastructure/release-schedule-versioning.rst [hunk 9]
```diff
@@ -74,21 +83,23 @@ Examples
...
-| ``0.4.0``     | Major marketing release                         |
-+---------------+-------------------------------------------------+
++----------------+-------------------------------------------------+
+| ``0.3.0``      | 0.3.0 released                                  |
++----------------+-------------------------------------------------+
+| ``0.3.1dev1``  | Weekly releases of 0.3.1                        |
++----------------+-------------------------------------------------+
+| ``0.3.1``      | Micro marketing release                         |
++----------------+-------------------------------------------------+
+| ``0.3.1+doc1`` | Documentation release of 0.3.1                  |
```
Comment (adamtheturtle) [inline docs/gettinginvolved/infrastructure/release-schedule-versioning.rst]: `+doc1` is missing the `.` from `+doc.X`. On that note, how come this is the only one with a period?

File: flocker/docs/_version.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+_VERSION_RE = re.compile(
+    # The base version (in 'release')
+    r"(?P<release>[0-9]\.[0-9]+\.[0-9]+)"
+    # For a documentation release, the release number.
+    # (in 'doc-release').
+    r"(\+doc\.(?P<doc>[0-9]+))?"
+    # For development version, the number of commits since the last
+    # release and git hash
+    r"((?P<development>-[0-9]+-g[0-9a-f]+))?"
+    # Wether the tree is dirty.
```
Comment (adamtheturtle) [inline flocker/docs/_version.py]: Typo s/Wether/whether

File: flocker/docs/_version.py (hunk 1 above)
Comment (adamtheturtle) [inline flocker/docs/_version.py]: I guess as this seems to work and it is not critical code, the current is fine.

File: docs/gettinginvolved/infrastructure/release-process.rst [hunk 11]
```diff
@@ -325,18 +331,24 @@ Release
...
 
+   .. note:: Skip this step for a documentation release.
+
    .. code-block:: console
 
       admin/upload-rpms "${VERSION}"
 
 #. Build and upload the tutorial :ref:`Vagrant box <build-vagrant-box>`.
 
+   .. note:: Skip this step for a documentation release.
```
Comment (adamtheturtle) [inline docs/gettinginvolved/infrastructure/release-process.rst]: This one means that the "Access" at the top to "Vagrant Cloud" is not necessary. Note also (make an issue?) that that link is also old now that Atlas is a thing.

File: docs/gettinginvolved/infrastructure/release-process.rst [hunk 12]
```diff
@@ -325,18 +331,24 @@ Release
 
 #. Build RPM packages and upload them to ``archive.clusterhq.com``
 
+   .. note:: Skip this step for a documentation release.
```
Comment (adamtheturtle) [inline docs/gettinginvolved/infrastructure/release-process.rst]: I think that this means that the gsutil "Access" at the top of this document is not necessary too.

**Comment by adamtheturtle** [comment]
Update the "Outcomes" section of the release process to say what will happen if doing a documentation release.

File: flocker/docs/test/test_version.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        documentation version is left unchanged.
+        """
+        self.assertParsedVersion('0.3.2+doc.11-dirty',
+                                 documentation_revision='11',
+                                 dirty='-dirty')
+
+
+class GetDocVersionTests(SynchronousTestCase):
+    """
+    Test for :function:`get_doc_version`.
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: s/Test/Tests

File: flocker/docs/test/test_version.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+        When the version is from a documentation release but is dirty, the
+        documentation version is left unchanged.
+        """
+        self.assertEqual(get_doc_version('0.3.2+doc.0-dirty'),
+                         '0.3.2+doc.0-dirty')
+
+
+class IsReleaseTests(SynchronousTestCase):
+    """
+    Test for :function:`is_release`.
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: s/Test/Tests

File: flocker/docs/test/test_version.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Test for :module:`flocker.docs.version`.
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: s/Test/Tests

File: flocker/docs/test/test_version.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+from .._version import (
+    parse_version, FlockerVersion,
+    get_doc_version, is_release,
+)
+
+
+class ParseVersionTests(SynchronousTestCase):
+    """
+    Test for :function:`parse_version`.
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: s/Test/Tests

File: flocker/docs/_version.py (hunk 3 above)
Comment (adamtheturtle) [inline flocker/docs/_version.py]: If you think that it is appropriate, create an issue to use the same regex in Buildbot.

File: flocker/docs/test/test_version.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+from twisted.trial.unittest import SynchronousTestCase
+
+from .._version import (
+    parse_version, FlockerVersion,
+    get_doc_version, is_release,
+)
+
+
+class ParseVersionTests(SynchronousTestCase):
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: Add tests (or create an issue for adding tests) for the other types of release (marketing, pre)

**Comment by adamtheturtle** [comment]
Thanks @tomprince , this looks good. I won't give the final go ahead because I'm very tired now, sorry, but I'll look over this tomorrow morning. If you're blocked on it, please hand the final review off to someone else.

File: flocker/docs/_version.py [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    A version of Flocker.
+
+    :ivar str major: The major number of the (most recent) release.
+    :ivar str minor: The minor number of the (most recent) release.
+    :ivar str micro: The micro number of the (most recent) release.
+    :ivar str documentation_revision: The documentation revision of the
+        (most recent) release or ``None`` if there hasn't been a documentation
+        release..
+    :ivar str commit_count: The number of commits since the last release or
+        ``None if this is a release.
```
Comment (adamtheturtle) [inline flocker/docs/_version.py]: Match "``" on "``None``"

File: docs/gettinginvolved/infrastructure/release-process.rst (hunk 11 above)
Comment (tomprince) [inline docs/gettinginvolved/infrastructure/release-process.rst]: https://clusterhq.atlassian.net/browse/FLOC-1289

File: docs/gettinginvolved/infrastructure/release-schedule-versioning.rst (hunk 9 above)
Comment (tomprince) [inline docs/gettinginvolved/infrastructure/release-schedule-versioning.rst]: No good reason.

**Comment by tomprince** [comment]
I believe I've addressed all the comments to this point.

File: docs/gettinginvolved/infrastructure/release-schedule-versioning.rst [hunk 19]
```diff
@@ -53,6 +53,15 @@ Minor marketing releases will be planned and scheduled by ClusterHQ's product te
 
 The version of a minor marketing release will have the micro version number incremented from the previous marketing release.
 
+Documentation Release
+^^^^^^^^^^^^^^^^^^^^^
+Documentation releases will be made when documentation for a major or minor marketing release is to be updated, without doing a full release.
+
+Documentation releases will be planned and scheduled by ClusterHQ's product team, in consultation with the marketing and engineering teams.
+
+The version of a documentation will have the version of the corresponding marketing release, with a ``+doc.X`` release, where ``X`` starts at ``1`` and is incremented for each documentation release.
```
Comment (adamtheturtle) [inline docs/gettinginvolved/infrastructure/release-schedule-versioning.rst]: +doc.X > +docX

**Comment by adamtheturtle** [comment]
Add `__init__.py` to `flocker.docs.test`.

**Comment by adamtheturtle** [comment]
Address the two comments, confirm that the new tests are running on Buildbot and merge when Buildbot passes to your satisfaction.

File: flocker/docs/test/test_version.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        When the documentation version is from a documentation release, it is a
+        release.  """
+        self.assertTrue(is_release('0.3.2+doc11'))
+
+    def test_doc_dirty(self):
+        """
+        When the version is from a documentation release but is dirty, it isn't
+        a release.
+        """
+        self.assertFalse(is_release('0.3.2+doc0-dirty'))
```
Comment (adamtheturtle) [inline flocker/docs/test/test_version.py]: :question: The incrementing doc number starts at 1 so this (and the other occurrences of doc0) is not a realistic case.

File: flocker/docs/__init__.py [hunk 21]
```diff
@@ -0,0 +1,5 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
```
Comment (adamtheturtle) [inline flocker/docs/__init__.py]: Docstring

**Comment by adamtheturtle** [comment]
I have addressed the remaining comments.

**Comment by adamtheturtle** [comment]
I can see the tests running on the ubuntu-14.04 builder for example.
